### ISSUE TITLE
Errors running on OSX

### ISSUE THREAD

**Issue by alohawolf** [topkecleon/telegram-bot-bash#122]
So now that I've fixed the environment issues, I'm running into another issue where it does not correctly parse the return from the API, see below for me running it in debug mode

kowela:telegram-bot-bash-0.94 aloha$ bash -x bashbot.sh start
+ '[' -t 1 ']'
+ '[' -n xterm-256color ']'
+ CLEAR=clear
+ RED='\e[31m'
+ GREEN='\e[32m'
+ ORANGE='\e[35m'
+ NC='\e[0m'
+ SCRIPT=bashbot.sh
+ REALME=bashbot.sh
++ dirname bashbot.sh
+ SCRIPTDIR=.
++ dirname bashbot.sh
+ RUNDIR=.
+ MODULEDIR=./modules
+ '[' bashbot.sh '!=' bashbot.sh ']'
+ '[' start = source ']'
++ basename bashbot.sh
+ SCRIPT=./bashbot.sh
++ basename ./modules
+ MODULEDIR=./modules
+ '[' -n '' ']'
+ ADDONDIR=./addons
+ RUNUSER=aloha
+ '[' '' '!=' yes ']'
+ '[' -z '' ']'
+ cd .
+ RUNDIR=.
+ '[' '!' -w . ']'
+ TOKENFILE=./token
+ BOTADMIN=./botadmin
+ BOTACL=./botacl
+ DATADIR=./data-bot-bash
+ COUNTFILE=./count
+ '[' -z '' ']'
+ '[' '!' -f ./token ']'
+ '[' -z '' ']'
+ BOTTOKEN= - - 
+ '[' '!' -f ./botadmin ']'
+ '[' '!' -f ./botacl ']'
+ '[' '!' -d ./data-bot-bash ']'
+ '[' '!' -w ./data-bot-bash ']'
+ '[' '!' -f ./count ']'
+ '[' '!' -w ./count ']'
+ URL=https://api.telegram.org/- - 
+ ME_URL=https://api.telegram.org/- - /getMe
+ UPD_URL='https://api.telegram.org/- - /getUpdates?offset='
+ GETFILE_URL=https://api.telegram.org/- - /getFile
+ declare -rx SCRIPT SCRIPTDIR MODULEDIR RUNDIR ADDONDIR TOKENFILE BOTADMIN BOTACL DATADIR COUNTFILE
+ declare -rx BOTTOKEN URL ME_URL UPD_URL GETFILE_URL
+ declare -ax CMD
+ declare -Ax UPD BOTSENT USER MESSAGE URLS CONTACT LOCATION CHAT FORWARD REPLYTO VENUE iQUERY SERVICE NEWMEMBER
+ export res CAPTION
+ COMMANDS=./commands.sh
+ '[' '' '!=' yes ']'
+ '[' '!' -f ./commands.sh ']'
+ '[' '!' -r ./commands.sh ']'
+ source ./commands.sh source
++ export LC_ALL=C.UTF-8
++ LC_ALL=C.UTF-8
++ export LANG=C.UTF-8
++ LANG=C.UTF-8
++ export LANGUAGE=C.UTF-8
++ LANGUAGE=C.UTF-8
++ unset IFS
++ bashbot_info='This is bashbot, the Telegram bot written entirely in bash.
It features background tasks and interactive chats, and can serve as an interface for CLI programs.
It currently can send, recieve and forward messages, custom keyboards, photos, audio, voice, documents, locations and video files.
'
++ bashbot_help='*Available commands*:
*• /start*: _Start bot and get this message_.
*• /help*: _Get this message_.
*• /info*: _Get shorter info message about this bot_.
*• /question*: _Start interactive chat_.
*• /cancel*: _Cancel any currently running interactive chats_.
*• /kickme*: _You will be autokicked from the chat_.
*• /leavechat*: _The bot will leave the group with this command _.
Written by Drew (@topkecleon), Daniil Gentili (@danogentili) and KayM(@gnadelwartz).
Get the code in my [GitHub](http://github.com/topkecleon/telegram-bot-bash)
'
++ '[' -n source ']'
++ for modules in "${MODULEDIR:-.}"/*.sh
++ [[ source == *\d\e\b\u\g* ]]
+++ basename ./modules/aliases.sh
++ _is_function aliases.sh
++ true
++ for modules in "${MODULEDIR:-.}"/*.sh
++ [[ source == *\d\e\b\u\g* ]]
+++ basename ./modules/answerInline.sh
++ _is_function answerInline.sh
++ true
++ for modules in "${MODULEDIR:-.}"/*.sh
++ [[ source == *\d\e\b\u\g* ]]
+++ basename ./modules/background.sh
++ _is_function background.sh
++ true
++ for modules in "${MODULEDIR:-.}"/*.sh
++ [[ source == *\d\e\b\u\g* ]]
+++ basename ./modules/chatMember.sh
++ _is_function chatMember.sh
++ true
++ for modules in "${MODULEDIR:-.}"/*.sh
++ [[ source == *\d\e\b\u\g* ]]
+++ basename ./modules/jsonDB.sh
++ _is_function jsonDB.sh
++ true
++ for modules in "${MODULEDIR:-.}"/*.sh
++ [[ source == *\d\e\b\u\g* ]]
+++ basename ./modules/sendMessage.sh
++ _is_function sendMessage.sh
++ true
++ export INLINE=0
++ INLINE=0
++ export 'FILE_REGEX=/.*'
++ FILE_REGEX='/.*'
++ '[' -r ./mycommands.sh ']'
++ source ./mycommands.sh source
+++ export res=
+++ res=
+++ export INLINE=0
+++ INLINE=0
+++ export 'FILE_REGEX=/.*'
+++ FILE_REGEX='/.*'
+++ export SILENCER=no
+++ SILENCER=no
+++ NOTADMIN='Sorry, this command is allowed for admin or owner only'
+++ NOTBOTADMIN='Sorry, this command is allowed for bot owner only'
+++ '[' source = startbot ']'
+++ '[' -f .mystartup ']'
++ '[' -z source ']'
++ [[ source == *\d\e\b\u\g* ]]
+ for modules in "${MODULEDIR:-.}"/*.sh
++ basename ./modules/aliases.sh
+ _is_function aliases.sh
+ true
+ for modules in "${MODULEDIR:-.}"/*.sh
++ basename ./modules/answerInline.sh
+ _is_function answerInline.sh
+ true
+ for modules in "${MODULEDIR:-.}"/*.sh
++ basename ./modules/background.sh
+ _is_function background.sh
+ true
+ for modules in "${MODULEDIR:-.}"/*.sh
++ basename ./modules/chatMember.sh
+ _is_function chatMember.sh
+ true
+ for modules in "${MODULEDIR:-.}"/*.sh
++ basename ./modules/jsonDB.sh
+ _is_function jsonDB.sh
+ true
+ for modules in "${MODULEDIR:-.}"/*.sh
++ basename ./modules/sendMessage.sh
+ _is_function sendMessage.sh
+ true
+ declare -xr DELETE_URL=https://api.telegram.org/- - /deleteMessage
+ TIMEOUT=
+ [[ '' =~ ^[0-9]+$ ]]
+ TIMEOUT=20
+ '[' -z '' ']'
+ _exists curl
+ true
+ declare -Ax BASBOT_EVENT_INLINE BASBOT_EVENT_MESSAGE BASHBOT_EVENT_CMD BASBOT_EVENT_REPLY BASBOT_EVENT_FORWARD BASHBOT_EVENT_SEND
+ declare -Ax BASBOT_EVENT_CONTACT BASBOT_EVENT_LOCATION BASBOT_EVENT_FILE BASHBOT_EVENT_TEXT BASHBOT_EVENT_TIMER
+ EVENT_SEND=0
+ EVENT_TIMER=0
+ _is_function send_message
+ true
+ JSONSHFILE=./JSON.sh/JSON.sh
+ [[ ./JSON.sh/JSON.sh != *\/\J\S\O\N\.\s\h ]]
+ '[' '!' -f ./JSON.sh/JSON.sh ']'
+ '[' '' '!=' yes ']'
+ '[' start '!=' init ']'
+ '[' start '!=' help ']'
++ getBotName
++ getJson https://api.telegram.org/- - /getMe
++ curl -sL -k -m 20 https://api.telegram.org/- - /getMe
++ ./JSON.sh/JSON.sh -s -b -n
++ JsonGetString '"result","username"'
++ sed -n -e '0,/\["result","username"\]/ s/\["result","username"\][ \t]"\(.*\)"$/\1/p'
+ ME=
+ '[' -z '' ']'
+ echo -e '\e[31mERROR: Can'\''t connect to Telegram! Your TOKEN is invalid or you are blocked by https://api.telegram.org ...\e[0m'
ERROR: Can't connect to Telegram! Your TOKEN is invalid or you are blocked by https://api.telegram.org ...
+ case "$1" in
+ exit 1

JSON results when I run the curl command manually. (also sanitized the ID)

{"ok":true,"result":{"id":1153069062,"is_bot":true,"first_name":"Aloha Test Bot","username":"AlohaTestBot","can_join_groups":true,"can_read_all_group_messages":true,"supports_inline_queries":false}}

**Comment by gnadelwartz** [comment]
uh# possibly to less coffee, but can you point me exactly to the error / problem? to much debug output and not able to see the wrong values / error

**Comment by gnadelwartz** [comment]
ok, you removed you bottoken, but is the bottoken in the line

curl -sL -k -m 20 https://api.telegram.org/- - /getMe

really the same? this must always breturn true with the correct token

**Comment by gnadelwartz** [comment]
whats the result of:

`curl "https://api.telegram.org/bot$(<token)/getMe"`

if your token in the file `token` is correct the answer should something like

`{"ok":true,"result":{"id":796814662,"is_bot":true," ... }`

if this is not the case then:

- delete the file token and enter it new with vi, vim, emacs or whatever cli editor, but
not with a Mac Editor, as they may not enter enter the correct line ending to the file.
- execute `unset BOTTOKEN` ans try again

**Comment by alohawolf** [comment]
For what it's worth my token works on Linux, also you can see my results from running the curl command manually at the bottom.

modern macos uses the same line ending conventions as unix, this has been true for the entire history of OSX.

**Comment by gnadelwartz** [comment]
sorry, I can't help you if you if you don't answer my questions

whats the result of:

`curl "https://api.telegram.org/bot$(<token)/getMe"`

**Comment by alohawolf** [comment]
I posted it on my original report:

JSON results when I run the curl command manually. (also sanitized the ID)

{"ok":true,"result":{"id":1153069062,"is_bot":true,"first_name":"Aloha Test Bot","username":"AlohaTestBot","can_join_groups":true,"can_read_all_group_messages":true,"supports_inline_queries":false}}

**Comment by gnadelwartz** [comment]
and this was exactly the command I wrote you with the read of the tokenfile?

this part: `"$(<token)"` is not a placeholder for type yor token here, it read the token from the file token while executing, excactly like its done by the script.

If you have done it this way already in your first post, then apologies ..

**Comment by alohawolf** [comment]
yes, it was, I actually copied the curl command from the debug log (which includes the full URL string, as I was suspecting a curl issue) and ran that outside the script, and got back the json which I pasted.

**Comment by gnadelwartz** [comment]
this is a usesless discussion ...

you don't follow my advises to debug your problem, not a clear instruction to
copy and exceut the command exactly as I wrote to you.

so the only answer I can give to you: sorry can't help.

**Comment by alohawolf** [comment]
This is the command I ran, curl -sL -k -m 20 https://api.telegram.org/- - /getMe with - - replaced by the API Key

this was the result I got:

{"ok":true,"result":{"id":1153069062,"is_bot":true,"first_name":"Aloha Test Bot","username":"AlohaTestBot","can_join_groups":true,"can_read_all_group_messages":true,"supports_inline_queries":false}}

Is this different than the command you wanted run?

**Comment by alohawolf** [comment]
```kowela:telegram-bot-bash-0.94 aloha$ curl -sL -k -m 20 https://api.telegram.org/bot- -/getMe
{"ok":true,"result":{"id":1153069062,"is_bot":true,"first_name":"Aloha Test Bot","username":"AlohaTestBot","can_join_groups":true,"can_read_all_group_messages":true,"supports_inline_queries":false}}
kowela:telegram-bot-bash-0.94 aloha$ cat token 
- -
kowela:telegram-bot-bash-0.94 aloha$ ```

**Comment by gnadelwartz** [comment]
If you exece the command I wrote to you above then your api token us not visible on the command line and there is no need to remove it. this is why I think you did not follow my advises.

nevertheless I have no access to mac nor an idea what's wrong.
did you already try an older version in case there is an error in newer versions?

@topkecleon do you have an Idea? i have to give up.

**Comment by gnadelwartz** [comment]
I added some BOTTOKEN checks with explanations when failed to dev branch, may be this helps if problems like this arises.

**Comment by alohawolf** [comment]
Here is the error I wget on linux in the develop branch:
```aloha@leho ~/telegram-bot-bash> ./bashbot.sh start
Warning, something may wrong with your token file.
The file musst be 1 newline,  1 word,  and  46 byte, your's is: 1 1 47
Session Name: AlohaTestBot-startbot
Bot started successfully.```

this is the errors I get on MacOS:
```kowela:telegram-bot-bash aloha$ ./bashbot.sh start
Warning, something may wrong with your token file.
The file musst be 1 newline,  1 word,  and  46 byte, your's is:       1       1      47
ERROR: Can't connect to Telegram! Your TOKEN is invalid or you are blocked by https://api.telegram.org ...```

**Comment by gnadelwartz** [comment]
a bot token looks like `123456789:xyz-2....`, mine and all i know consists if 9 digit's and 35 alunm character, how many digits in front of `:` have yours? if possible check how many characters after  `:` it have.

regarding printf, this should be a bash built-in IMHO

**Comment by gnadelwartz** [comment]
> I can provide access to this on a BSD machine or a shell account on a mac box, if needed.

this would be helpful, so I can do some tests, I would prefer Mac

**Comment by alohawolf** [comment]
mine is 10 numbers a : and then 35 letters.

it may take me a day or two, but I'll get you setup on it, I'll send you creds once I have it setup. I'll also turn up a BSD vm for you.

Also, printf in both cases is a binary, on linux its part of coreutils:

```kowela:telegram-bot-bash aloha$ whereis printf
/usr/bin/printf

aloha@leho ~/telegram-bot-bash> whereis printf
printf: /usr/bin/printf /usr/include/printf.h /usr/share/man/man1/printf.1.gz /usr/share/man/man3/printf.3.gz```

**Comment by gnadelwartz** [comment]
https://www.linuxjournal.com/content/bashs-built-printf-function

**Comment by gnadelwartz** [comment]
from thinking about (possible) differences in the core commands between (gnu) Linux versions and mac/bsd versions it may be easier to install the needed gnu variants and add them to the path ...

https://apple.stackexchange.com/questions/69223/how-to-replace-mac-os-x-utilities-with-gnu-core-utilities

**Comment by gnadelwartz** [comment]
does  the following line output BSD on mac and BSD with standard sed?

``` bash
 sed '1ia' </dev/null 2>/dec/null ||  echo 'BSD'
```

**Comment by gnadelwartz** [comment]
@alohawolf anything new on MacOS and BSD?
